toolversions/toolversions.go

```diff
@@ -0,0 +1,122 @@
+package toolversions
+
+import (
+	"fmt"
+	"path/filepath"
+	"regexp"
+	"strings"
+
+	"github.com/bitrise-io/go-utils/v2/command"
+	"github.com/bitrise-io/go-utils/v2/env"
+	"github.com/bitrise-io/go-utils/v2/log"
+)
+
+type ToolVersionReporter interface {
+	// IsAvailable returns true if the tool version manager is available and actively manages the tool versions.
+	IsAvailable() bool
+
+	// CurrentToolVersions returns a snapshot of the currently active tools and versions.
+	// The returned map is keyed by tool name.
+	// Tool names and reported versions are implementation-specific. Tool names are normalized to lowercase.
+	CurrentToolVersions() (map[string]ToolVersion, error)
+}
+
+type ToolVersion struct {
+	Version        string
+	IsInstalled    bool
+	DeclaredByFile string
+	IsGlobal       bool
+}
+
+type ASDFVersionReporter struct {
+	cmdLocator  env.CommandLocator
+	cmdFactory  command.Factory
+	logger      log.Logger
+	userHomeDir string
+}
+
+func NewASDFVersionReporter(cmdLocator env.CommandLocator, cmdFactory command.Factory, logger log.Logger, userHomeDir string) ASDFVersionReporter {
+	return ASDFVersionReporter{
+		cmdLocator:  cmdLocator,
+		cmdFactory:  cmdFactory,
+		logger:      logger,
+		userHomeDir: userHomeDir,
+	}
+}
+
+func (r *ASDFVersionReporter) IsAvailable() bool {
+	_, err := r.cmdLocator.LookPath("asdf")
+	if err != nil {
+		r.logger.Debugf("asdf not found in path")
+		return false
+	}
+
+	code, err := r.cmdFactory.Create("asdf", []string{"current"}, &command.Opts{}).RunAndReturnExitCode()
+	if err != nil {
+		r.logger.Debugf("run asdf current: %s", err)
+		return false
+	}
+	if code != 0 {
+		r.logger.Debugf("run asdf current: nonzero exit code: %d", code)
+		return false
+	}
+
+	return true
+}
+
+func (r *ASDFVersionReporter) CurrentToolVersions() (map[string]ToolVersion, error) {
+	cmd := r.cmdFactory.Create("asdf", []string{"current"}, &command.Opts{})
```

Comment:
what happens to default tools (e.g. Flutter) that aren't set by `asdf` but available in our stack? Do we capture them ?

Reply:
Not yet, but we are constantly trying to move more and more tools under ASDF (or any other version manager in the future).
But @godrei had a good analysis of Flutter versions recently that you might find interesting: https://bitrise.atlassian.net/wiki/spaces/ENGI/pages/2734293197/RFC+Flutter+SDK+version+on+stacks